### PR TITLE
fix None-type not iterable error when deepspeed is left blank w/ use_…

### PR DESCRIPTION
**Pull request by bursteratom**
If you use qlora and set `use_reentrant: false` while leaving deepspeed blank `deepspeed:  ` in the config yml, you get:

 `File "/workspace/axolotl/src/axolotl/utils/config/models/input/v0_4_1/__init__.py", line 1317, in warn_qlora_zero3_w_use_reentrant
    and "zero3" in data.get("deepspeed", "")
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: argument of type 'NoneType' is not iterable`

This PR fixes it.

# Description

### CODE REVIEW

File: src/axolotl/utils/config/models/input/v0_4_1/__init__.py [hunk 1]
```diff
@@ -1314,6 +1314,7 @@ def warn_qlora_zero3_w_use_reentrant(cls, data):
             and data.get("gradient_checkpointing_kwargs", {})
             and data.get("gradient_checkpointing_kwargs", {}).get("use_reentrant")
             is False
+            and data.get("deepspeed", "") is not None
             and "zero3" in data.get("deepspeed", "")
```
Comment (winglian) [inline src/axolotl/utils/config/models/input/v0_4_1/__init__.py:1318]: isn't this redundant with the next line? if it's empty, doesn't the next line `data.get("deepspeed", "")` default it to an empty string?

Comment (bursteratom) [reply src/axolotl/utils/config/models/input/v0_4_1/__init__.py:1318]: when deepspeed is not specified, `data.get("deepspeed", "")` returns `None` and I get `TypeError: argument of type 'NoneType' is not iterable`

**Comment by winglian** [comment]
let's add a unit test for this. Doesn't need to be an e2e test, just a basic config validation for the two cases where deepspeed  is empty vs not set in the config.

**Comment by winglian** [comment]
also, you can skip the E2E tests since they are not needed for this change. simply append `[skip e2e]` to your commit messages.

**Comment by bursteratom** [comment]
@winglian just added unit tests tho I'm not sure if I did it correctly as there doesn't seem to be room to generalise beyond the very specific combination `adapter: qlora` + `deepspeed: ` + `use_reentrant: false` . How should I modify the tests?

**Comment by winglian** [comment]
Look further down in that test validation file, like around line 200 in this pr. There is an example of checking the warning message. But also good to have the checks where it passes too.

**Review by winglian — APPROVED** [review]
